### ISSUE TITLE
[ramda] Improve `assoc` type definition

### ISSUE THREAD

**Issue by matsjj** [flowtype/flow-typed#2010]
This seems to be a fairly easy function, and I'm sure that flow could infer the resulting type almost automatically, if you were to write it out in idiomatic JavaScript. 

In a similar manner, I changed `merge` locally to 

```
  declare type Merge = <A, B>(A, B) => $Exact<{...A, ...B}>
```

to reflect the fact that B has overwritten/added values into A. I'm not sure on the exact, ideally this returns an Exact object if `A` was exact, otherwise it wouldn't. But I don't think that's possible currently. 

Coming back to `assoc`, currently it is

```
  declare function assoc<T, S>(
    key: string,
    val: T,
    src: S
  ): { [k: string]: T } & S;
```

which adds very little type information. 

Is there any way to use `key` as a key in an object definition here? Ideally I would do something like

```
{...{ key: T } ,...S};
```

but that would just add T with the literal `key`, instead of using the often constant. 
Any easy way to implement this?

**Comment by AndrewSouthpaw** [comment]
cc @LoganBarnett 

I'm pretty sure we'd want to use [`$ElementType`](https://flow.org/en/docs/types/utilities/#toc-elementtype), I just haven't updated it yet. Then we could get rid of the nonsensical premise that the object being `assoc`ed must have the same type for all its keys.

**Comment by LoganBarnett** [comment]
Agreed that the problem here is that it has to be a mapping object. Although I'm not sure how we can leverage `$ElementType` since we're adding a key, not referring to an existing one. I would think here we would want some kind of `$Literal` if such a thing could exist.

``` javascript
declare function assoc<K: $Literal<string>, V, S>(
    key: K,
    val: V,
    src: S,
  ): S & { K: V };
```

I've been kicking around the idea of `$Literal` for a bit in my head but couldn't come up with a concrete example. Maybe this is it?

**Comment by LoganBarnett** [comment]
@matsjj relating to your earlier comment: I don't believe it's sound to assume an exact type after a `merge`, since any inexact properties will have come along for the ride as well. The output should have the same exactness of the inputs, if that makes sense. Perhaps we should talk about it as a separate issue, but if you feel we should address exactness here let's figure out what's driving that.

**Comment by AndrewSouthpaw** [comment]
Derp, you're totally right, `$ElementType` does us no favors. Clearly I was writing that late at night.

**Comment by AndrewSouthpaw** [comment]
What is the purpose of the `$Literal` here?

**Comment by LoganBarnett** [comment]
`$Literal` as I imagine it would be a utility type that represents some literal value. Take this:

``` javascript
// @flow

type Foo = {
  type: 'foo',
  value: number,
}

type Bar = {
  type: 'bar',
  value: string,
}

type FooBar = Foo | Bar

const foobar = (fb: FooBar): number => {
  switch(fb) {
  case 'foo':
    return fb.value + 1
  case 'bar':
    return parseInt(fb.value)
  // just imagine Flow works with exhaustive checks for a moment
  }
}
```

Literal values as types in Flow can give us a lot of utility. Right here we have a union that uses a sentinel property `type` and each type in the union has a literal set for that type. I imagine you probably know about this already but I wanted to set a stage with it. I found out you can even use literals as a means of providing function overloads for things like the `.on('eventName', ...)` in what I think is `EventEmitter`. But when you go passing these things around, you have to use their base type instead. What if you could pass them around generically?

``` javascript
// @flow

type Foo = {
  foo: number,
}

const key = 'foo'

const foo: Foo = { [key]: 1 } // works as expected!

const getFooValue = (): string => 'foo'

const key2 = getFooValue()
const foo2: Foo = { [key2]: 2 } // oh noes!

const getFooValueLiteral = () => 'foo'

const key3 = getFooValueLiteral()
const foo3: Foo = { [key3]: 2 } // Works again
```

There's some special type information that's in there. In fact, this can be hardcoded as a type:

``` javascript
// @flow

const getFooValueLiteralExplicit = (): 'foo' => 'foo'

const key = getFooValueLiteralExplicit()
const foo: Foo = { [key]: 2 } // works
```

But what I would like to see is something that allows me to express something like `$Literal<string>` (constrained so I can limit what it is - or use `*` if I don't care), and pass them around generically. Flow seems to understand a literal differently from other things and gives it special treatment, but short of the actual value being typed, I can't use it generically.

In the case of `assoc`, `$Literal` would be the key and because we've enforced that it will be a literal, we should be able to carry it over as part of the type def for the intersection create with `S & { K: V }`. It knows that `K` is legitimately a type that can be a property, and at some point that exact property can be resolved as part of feeding all of the typed information to `assoc` during our invocation.

Does that kind of make sense?

**Comment by AndrewSouthpaw** [comment]
Nicely explained, that does make sense. :)

Now for the big question: do you know ML so you can implement it in Flow? 😋

**Comment by AndrewSouthpaw** [comment]
Nicely explained, that does make sense. :)

Now for the big question: do you know ML so you can implement it in Flow? 😋
